internal/initwd/module_install.go

```diff
@@ -405,15 +427,23 @@ func (i *ModuleInstaller) installRegistryModule(req *earlyconfig.ModuleRequest,
 	// If we manage to get down here then we've found a suitable version to
 	// install, so we need to ask the registry where we should download it from.
 	// The response to this is a go-getter-style address string.
-	dlAddr, err := reg.ModuleLocation(addr, latestMatch.String())
-	if err != nil {
-		log.Printf("[ERROR] %s from %s %s: %s", key, addr, latestMatch, err)
-		diags = diags.Append(tfdiags.Sourceless(
-			tfdiags.Error,
-			"Invalid response from remote module registry",
-			fmt.Sprintf("The remote registry at %s failed to return a download URL for %s %s.", hostname, addr, latestMatch),
-		))
-		return nil, nil, diags
+	var dlAddr string
+
+	// check if we've already looked up the module download URL
+	if dlAddr, exists = i.moduleVersionsUrl[moduleVersion{module: addr.String(), version: latestMatch.String()}]; exists {
```

Comment:
This bit made me double-take a little because I didn't grok initially that this is using the `dlAddr` declared above, and was instead expecting it to be more like the similar branching above where the condition is only deciding whether to populate the map, not also writing to the result variable.

This is totally subjective, so no worries if you disagree, but I might be more inclined to write it this way:

```go
moduleAddr := moduleVersion{module: addr.String(), version: latestMatch.String()}

if _, exists := i.moduleVersionsUrl[moduleAddr]; !exists {
    // do all the fetching logic, and then populate the cache map:
    i.moduleVersionsUrl[moduleAddr] = ...
}

// now dlAddr can be declared and assigned in the same place, regardless
// of whether we'd had it cached.
dlAddr := i.moduleVersionsUrl[moduleAddr]
```

Reply:
Looks great to me, thank you! I just pushed an update.